### ISSUE TITLE
How to disable replacing characters of an entry title from SBC case To DBC case ?

### ISSUE THREAD

**Issue by peng-howard** [retorquere/zotero-better-bibtex#657]
When exporting BibTeX, better bibtex replaces SBC case characters to DBC case ones, that is, titles such as "some words：the other，words" in zotero will be replaced by titles such as "some words:the other,words" after zotero entries are exported to a .bib file

Besides titles, path or file name of a attachment file is also processed by the replacing rules and I can't open the attachment file of a entry in Jabref with the exported .bib file just because the real path of a file（"fileNamePart1：fileNamePart2.pdf"） is now given by a wrong path（"fileNamePart1:fileNamePart2.pdf"），so is there a way to sovle this? thx.

**Comment by retorquere** [comment]
A few things:

1. I have no idea what SBC and DBC means.
2. Can you submit an error report on a specific reference withe a title like `some words：the other，words` by right-clicking it and selecting "submit better bibtex error"? If you post the debug ID you get after submission here I'll know how to get it.
3. the path handling is a different issue, can you submit it in a separate ticket? Easier for me to track.

**Comment by peng-howard** [comment]
1. In a Chinese title, we always use `：` instead of `:`,  and we also use `，` as a substitution of  `,`. We call characters such as `：，（）。《》“‘’”` Double Byte Characters (DBC), and call characters `:,().<>` Single Byte Characters(SBC).

2. For a Chinese entry titled by `流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究` together with an attachment file `流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究.pdf` in Zotero, the result `.bib` file(BibTeX) exported by Better BibTeX is:

```
@article{SuDongWei2004Pricing,
  title = {{流动性与资产定价:基于我国股市资产换手率与预期收益的实证研究}},
  volume = {2},
  shorttitle = {{流动性与资产定价}},
  language = {Chinese},
  journal = {经济研究},
  author = {{苏冬蔚} and {麦元勋}},
  year = {2004},
  pages = {95--105},
  file = {流动性与资产定价:基于我国股市资产换手率与预期收益的实证研究,苏冬蔚,2004,经济研究.pdf:D\:\\OneDrive\\storage\\流动性与资产定价:基于我国股市资产换手率与预期收益的实证研究,苏冬蔚,2004,经济研究.pdf:application/pdf}
}
```

3. As is shown in the above code, **title** field's value is now `流动性与资产定价:基于我国股市资产换手率与预期收益的实证研究`, `：` in the original title is replaced by `:`. In fact, we just need a title likes `流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究`

4. The real file name `流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究.pdf` is replaced by `流动性与资产定价:基于我国股市资产换手率与预期收益的实证研究.pdf` and Jabref told us the attachment file `流动性与资产定价:基于我国股市资产换手率与预期收益的实证研究.pdf` cannot be found. Though there is no error reported by Better BibTeX, it has done too much than what was expected.

Thx.

**Comment by retorquere** [comment]
1. OK
2. I need a source reference for diagnosis. If you right-click the relevant reference(s) and then select "submit better bibtex error" from the popup menu you will get a dialog allowing to submit an error report. After submission you will get a debug ID in red; if you post that debug ID here I can diagnose the problem.
3. I can only diagnose this after I have the source reference, submitted as in 2.
4. please file a separate issue, because attachment path handling is an entirely different bit of code and I prefer to track distinct issues in separate issue tickets.

**Comment by peng-howard** [comment]
I have submitted the debug report and the debug ID is `ERQGMCX3`.

**Comment by retorquere** [comment]
The reference you submitted can be found at https://gist.github.com/3b33d9b83e03f3fc744510b9fd34f629; can you see if the reference has DBC or SBC? How would I be able to tell?

**Comment by duncdrum** [comment]
@retorquere you can see it in the spacing:
 single byte space (U+003A)```:字```   
double byte space (U+FF1A)```：字```,  
both inlines consist of only 2 unicode glyphs.

**Comment by retorquere** [comment]
OK, but is the JSON file at that link on DBC or SBC? You wildly overestimate my knowledge of unicode.

**Comment by duncdrum** [comment]
@retorquere most of unicode is double byte just the old ascii stuff is single byte, so its not a per file property but a per character property. The JSON/SGML/CSV either uses U003A or UFF1A. the more curious case is why and where it would convert. I guess on legacy filesystems paths need to be ascii, but in 2017 assuming unicode compliance seems a reasonable assumption to me.

**Comment by retorquere** [comment]
OK so I understand that much :) but it's common to think of files as having a single encoding throughout (even if an encoding like UTF-8 has codepoints which share meaning with those in ASCII) -- things would get pretty crazy if you'd have to guess per-character, because there's (to my knowledge at least) no way in unicode to denote "next character is UTF-16".

In any case, what I'm trying to assess is whether BBT gets the DBC or SBC version -- Javascript has really, [*really*](https://mathiasbynens.be/notes/javascript-unicode) funky unicode handling, and I'd have to know whether things go wrong on my end (which is possible -- I do a normalization to NFKC before outputting latex code) or whether it's done by Zotero. I was hoping that the JSON export would give us enough info, as it doesn't have normalization applied, but then JSON always *will* be UTF-8, and I have no idea what the source encoding may have been.

That it's in the CSV export means it may be Zotero's doing, in which case there'd be nothing I can do about it. I just don't yet know how to find out.

**Comment by peng-howard** [comment]
I exported the entry by using `BibTeX` provided by zotero itself instead of Better BibTeX, now the result `.bib` file gives the right case:

```
@article{__2004,
	title = {流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究},
	volume = {2},
	shorttitle = {流动性与资产定价},
	url = {http://www.erj.cn/UploadFiles/%E6%B5%81%E5%8A%A8%E6%80%A7%E4%B8%8E%E8%B5%84%E4%BA%A7%E5%AE%9A%E4%BB%B7_%E5%9F%BA%E4%BA%8E%E6%88%91%E5%9B%BD%E8%82%A1%E5%B8%82%E8%B5%84%E4%BA%A7%E6%8D%A2%E6%89%8B%E7%8E%87%E4%B8%8E%E9%A2%84%E6%9C%9F%E6%94%B6%E7%9B%8A%E7%9A%84%E5%AE%9E%E8%AF%81%E7%A0%94%E7%A9%B6.pdf},
	language = {Chinese},
	urldate = {2015-03-30},
	journal = {经济研究},
	author = {{苏冬蔚} and {麦元勋}},
	year = {2004},
	note = {bibtex: SuDongWei2004Pricing},
	pages = {95--105},
	file = {流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究，苏冬蔚，2004，经济研究.pdf:D\:\\OneDrive\\storage\\+ 金融\\+ 专题\\流动性\\流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究，苏冬蔚，2004，经济研究.pdf:application/pdf}
}
```

And the above code shows that something goes wrong just because of Better BibTeX.

**Comment by retorquere** [comment]
That seems unambiguous then. Hmm. I'll go check.

**Comment by duncdrum** [comment]
How does the normalization pipeline work? JS indeed can be funky with unicode. But the DBC peng-howard is referring to are all UTF-8 no need for UTF-16 double-double-bytes.  
That being said: this (from the link you posted) is correct:
```JSON
"attachments": [
        {
          "path": "D:\\OneDrive\\storage\\+ 金融\\+ 专题\\流动性\\流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究，苏冬蔚，2004，经济研究.pdf",
          "title": "流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究，苏冬蔚，2004，经济研究.pdf",
          "mimeType": "application/pdf"
        }
],
```

**Comment by retorquere** [comment]
I don't know how it works, just how to [call it](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/normalize). I added it in early 2015 to get around some problems with my Travis tests.

Circle (which I use now) has no such issues however -- @peng-howard, could you try https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-1.6.95-br657-3702.xpi ? This has the unicode normalization turned off, maybe it's just that.

**Comment by duncdrum** [comment]
I m just worried that turning off unicode normalization could have unintended side effects, e.g. Author names with accents will now be different, pinyin transliteration will for sure. Are all fields of Zotero normalized before being processed by BBT, or do you call it on the something already generated by bbt? Id be happy to help you design some tests, but i don't quite understand when and where you normalize. Was it just because of travis? 
I ll have a closer look at cricle and see if I can come up with some test if you could use them. Just changing the normalisation form might be the less invasive way to go.

**Comment by retorquere** [comment]
Normalization has to do with compositing characters -- see [here](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/normalize). There are different ways to get the same composed character on screen, and normalization switches between these ways.

I don't know whether Zotero does any normalization but I don't think it does, or the standard bibtex export would have exhibited the same behavior. In any case, the sole reason I added it was to pacify Travis, no other reason, and I don't use Travis anymore.

**Comment by duncdrum** [comment]
Yes i know what unicode normalisation does. The OPs example doesn't contain any combined Latin characters, but i m wondering what happens if the entry has transliteration info for eg: author ```Sū Dōngwèi``` u, o, e will differ depending on normalization. I ll run some test with the branch, later tonight.

**Comment by retorquere** [comment]
For what it's worth, that test xpi has passed all my own tests (can't release anything that doesn't); I'd expect it to pass yours. For the same reason I don't fully expect the test xpi to fix OPs problem, but I would prefer to scratch the potential cause off the list.

**Comment by peng-howard** [comment]
@retorquere I have tested `BBT 3702`:
1. The entry I have used:
```
@article{SuDongWei2004Pricing,
  title = {{流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究}},
  volume = {2},
  shorttitle = {{流动性与资产定价}},
  language = {Chinese},
  timestamp = {2017-04-18T17:24:33Z},
  urldate = {2015-03-30},
  journal = {经济研究},
  author = {{苏冬蔚} and {麦元勋}},
  year = {2004},
  pages = {95--105},
  file = {流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究，苏冬蔚，2004，经济研究.pdf:D\:\\OneDrive\\storage\\+ 金融\\+ 专题\\流动性\\流动性与资产定价：基于我国股市资产换手率与预期收益的实证研究，苏冬蔚，2004，经济研究.pdf:application/pdf}
}
```

2. Author name with accents: `Holmström`
```
@article{Holmstrom1990Regulating,
  title = {Regulating Trade among Agents},
  timestamp = {2017-02-08T01:53:57Z},
  urldate = {2017-02-08},
  journal = {Journal of Institutional and Theoretical Economics},
  author = {Holmstr{\"o}m, Bengt and Milgrom, Paul},
  year = {1990},
  keywords = {多代理人},
  pages = {85--105},
  file = {Regulating Trade Among Agents, Holmström & Milgrom(1990), Journal of Institutional and Theoretical Economics.pdf:D\:\\OneDrive\\storage\\+ 经济\\+ 信息、合同、激励、机制设计\\Regulating Trade Among Agents, Holmström & Milgrom(1990), Journal of Institutional and Theoretical Economics.pdf:application/pdf}
}
```
@duncdrum  Note that in the second case, `author` field `Holmström` has a value of `Holmstr{\"o}m`, `bibkey` has a value of `Holmstrom`, and the value of `file` is `...Holmström...`. By the way, when managing our refs, `Sū Dōngwèi` is not the one we adopt in most cases, as you can see in bibkey  `SuDongWei2004Pricing`, `Su Dongwei` is the right one.

Yes, this is the exact one I have expected!

Hmm, A perfect `.bib` file.

**Comment by duncdrum** [comment]
@peng-howard yup citekey, and other data look good, thanks for checking. 

@retorquere looks like another extremely quick fix, great job.

**Comment by retorquere** [comment]
OK I understand about half of what's going on here but the bottom line seems to be that the two of you are happy :) I'll merge and cut a new release.

The paths in #658 are OK as-is? That sort of surprises me as I thought colons are spoil-sports in that field. But if it works, it works. Feel free to close #658 then.

**Comment by peng-howard** [comment]
Great job, I have closed #658 already. @retorquere @duncdrum

**Comment by retorquere** [comment]
Is this issue also OK to close?
